ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java

```diff
@@ -0,0 +1,476 @@
+        checkBlobId(get(), getOperationName());
+      }
+    };
+    opInfo.futures.add(testFuture);
+  }
+
+  private void startGetBlobInfo(boolean nonBlocking, final boolean afterDelete, final OperationInfo opInfo) {
+    Callback<BlobInfo> callback = new TestCallback<>(opInfo, afterDelete);
+    Router router = nonBlocking ? nonBlockingRouter : coordinatorBackedRouter;
+    Future<BlobInfo> future = router.getBlobInfo(opInfo.blobId, callback);
+    TestFuture<BlobInfo> testFuture =
+        new TestFuture<BlobInfo>(future, genLabel("getBlobInfo", nonBlocking, afterDelete), opInfo) {
+          @Override
+          void check() {
+            if (afterDelete) {
+              try {
+                future.get();
+                Assert.fail("Blob should have been deleted in operation: " + getOperationName());
+              } catch (Exception ignored) {
+              }
+            } else {
+              checkBlobInfo(get(), opInfo, getOperationName());
+            }
+          }
+        };
+    opInfo.futures.add(testFuture);
+  }
+
+  private void startGetBlob(boolean nonBlocking, final boolean afterDelete, final OperationInfo opInfo) {
+    Callback<ReadableStreamChannel> callback = new TestCallback<>(opInfo, afterDelete);
+    Router router = nonBlocking ? nonBlockingRouter : coordinatorBackedRouter;
+    Future<ReadableStreamChannel> future = router.getBlob(opInfo.blobId, callback);
+    TestFuture<ReadableStreamChannel> testFuture =
+        new TestFuture<ReadableStreamChannel>(future, genLabel("getBlob", nonBlocking, afterDelete), opInfo) {
+          @Override
+          void check() {
+            if (afterDelete) {
+              try {
+                future.get();
+                Assert.fail("Blob should have been deleted in operation: " + getOperationName());
+              } catch (Exception ignored) {
+              }
+            } else {
+              checkBlob(get(), opInfo, getOperationName());
+            }
+          }
+        };
+    opInfo.futures.add(testFuture);
+  }
+
+  private void startDeleteBlob(boolean nonBlocking, final OperationInfo opInfo) {
+    Callback<Void> callback = new TestCallback<>(opInfo, false);
+    Router router = nonBlocking ? nonBlockingRouter : coordinatorBackedRouter;
+    Future<Void> future = router.deleteBlob(opInfo.blobId, callback);
+    TestFuture<Void> testFuture = new TestFuture<Void>(future, genLabel("deleteBlob", nonBlocking, false), opInfo) {
+      @Override
+      void check() {
+        get();
+      }
+    };
+    opInfo.futures.add(testFuture);
+  }
+
+  private void startAwaitCreation(final OperationInfo opInfo) {
+    notificationSystem.awaitBlobCreations(opInfo.blobId);
+    continueChain(opInfo);
+  }
+
+  private void startAwaitDeletion(final OperationInfo opInfo) {
+    notificationSystem.awaitBlobDeletions(opInfo.blobId);
+    continueChain(opInfo);
+  }
+
+  private void continueChain(final OperationInfo opInfo) {
+    synchronized (opInfo.futures) {
```

Comment:
So, it's true that `continueChain` calls within a single operation chain are sequential.  However, consider the scenario where we are executing the last operation in a chain.  There is the slight possibility that this last operation can complete and call its callback before this callback has finished executing.  This means that the operation can be marked as completed (by flipping the latch) before the last operation's future was added to the list of futures.  This means that the check method might miss checking the last future in a chain.